### PR TITLE
Added: Request matching for jsonrpccommands

### PR DESCRIPTION
**Pull request by MarcelRoozekrans**
### Functional description

Added request matching to find the suited response for the jsonrpccommand methods. 

### Reasoning

Couldn't make settings.getSettingValue to work with current implementation

### Technical description

Technical description
- Find the file for the jsonrcpcommand method
- Match the request to the one inside the stub file
- Return the mathced response
- If there is no response we fallback to a basic Ok command return

### CODE REVIEW

**Comment by basrieter** [comment]
Look promising. Do you have an example for a response file?

File: xbmc.py [hunk 1]
```diff
@@ -39,13 +38,12 @@
...
 TRAY_OPEN = 16
 
-
 # Custom AddonData type
 AddonData = namedtuple('AddonData', [
-    'kodi_home_path',       # data path (either portable of user path)
-    'add_on_id',            # the add-on id
-    'add_on_path',          # the full path to the add-on
-    'kodi_profile_path'     # the full path to the add-on profile folder
+    'kodi_home_path',  # data path (either portable of user path)
```
Comment (basrieter) [inline xbmc.py]: Could you revert this change in comment alignment? Somehow my OCD-brain can't handle it :)

Comment (MarcelRoozekrans) [reply xbmc.py]: haha, my pycharm tried 4 times to realign it :) Last time I gave in

**Comment by MarcelRoozekrans** [comment]
> Look promising. Do you have an example for a response file?

See:
 
```
[{
        "request": {
            "params": {
                "setting": "network.usehttpproxy"
            },
            "jsonrpc": "2.0",
            "method": "Settings.GetSettingValue",
            "id": 0
        },
        "response": {
            "id": 5,
            "jsonrpc": "2.0",
            "result": {
                "value": false
            }
        }
    }
]
```

File: xbmc.py (hunk 1 above)
Comment (MarcelRoozekrans) [reply xbmc.py]: @basrieter updated

**Comment by basrieter** [comment]
> > Look promising. Do you have an example for a response file?
> 
> See:
> 
> ```
> [{
>         "request": {
>             "params": {
>                 "setting": "network.usehttpproxy"
>             },
>             "jsonrpc": "2.0",
>             "method": "Settings.GetSettingValue",
>             "id": 0
>         },
>         "response": {
>             "id": 5,
>             "jsonrpc": "2.0",
>             "result": {
>                 "value": false
>             }
>         }
>     }
> ]
> ```

Ok, that makes sense. I see two issue:

1. the "id" is the request is generated by Kodi. I would make your compare on just the `params` instead of the full `request`. 
1. There is still no way to just _ignore_ the `request` and return a specific response based on just the json RPC method name. For example:  I stub the `Textures.RemoveTexture` and always want the same response, regardless of the input. Can we add a catch-all the filtering of the responses? If no specific response is found, return the catch-all if there is one, else return the default on from code.

**Comment by MarcelRoozekrans** [comment]
> > > Look promising. Do you have an example for a response file?
> > 
> > 
> > See:
> > ```
> > [{
> >         "request": {
> >             "params": {
> >                 "setting": "network.usehttpproxy"
> >             },
> >             "jsonrpc": "2.0",
> >             "method": "Settings.GetSettingValue",
> >             "id": 0
> >         },
> >         "response": {
> >             "id": 5,
> >             "jsonrpc": "2.0",
> >             "result": {
> >                 "value": false
> >             }
> >         }
> >     }
> > ]
> > ```
> 
> Ok, that makes sense. I see two issue:
> 
> 1. the "id" is the request is generated by Kodi. I would make your compare on just the `params` instead of the full `request`.
> 2. There is still no way to just _ignore_ the `request` and return a specific response based on just the json RPC method name. For example:  I stub the `Textures.RemoveTexture` and always want the same response, regardless of the input. Can we add a catch-all the filtering of the responses? If no specific response is found, return the catch-all if there is one, else return the default on from code.

Funny, in my tests I always get id: 0 so I think the counter in saké is not working right or maybe not implemented. :)

**Comment by basrieter** [comment]
> > > > Look promising. Do you have an example for a response file?
> > > 
> > > 
> > > See:
> > > ```
> > > [{
> > >         "request": {
> > >             "params": {
> > >                 "setting": "network.usehttpproxy"
> > >             },
> > >             "jsonrpc": "2.0",
> > >             "method": "Settings.GetSettingValue",
> > >             "id": 0
> > >         },
> > >         "response": {
> > >             "id": 5,
> > >             "jsonrpc": "2.0",
> > >             "result": {
> > >                 "value": false
> > >             }
> > >         }
> > >     }
> > > ]
> > > ```
> > 
> > 
> > Ok, that makes sense. I see two issue:
> > 
> > 1. the "id" is the request is generated by Kodi. I would make your compare on just the `params` instead of the full `request`.
> > 2. There is still no way to just _ignore_ the `request` and return a specific response based on just the json RPC method name. For example:  I stub the `Textures.RemoveTexture` and always want the same response, regardless of the input. Can we add a catch-all the filtering of the responses? If no specific response is found, return the catch-all if there is one, else return the default on from code.
> 
> Funny, in my tests I always get id: 0 so I think the counter in saké is not working right or maybe not implemented. :)

I guess that is true. But still, any add-on could decide to make it > 0. The second item is more important, I think.

**Comment by MarcelRoozekrans** [comment]
> > > > > Look promising. Do you have an example for a response file?
> > > > 
> > > > 
> > > > See:
> > > > ```
> > > > [{
> > > >         "request": {
> > > >             "params": {
> > > >                 "setting": "network.usehttpproxy"
> > > >             },
> > > >             "jsonrpc": "2.0",
> > > >             "method": "Settings.GetSettingValue",
> > > >             "id": 0
> > > >         },
> > > >         "response": {
> > > >             "id": 5,
> > > >             "jsonrpc": "2.0",
> > > >             "result": {
> > > >                 "value": false
> > > >             }
> > > >         }
> > > >     }
> > > > ]
> > > > ```
> > > 
> > > 
> > > Ok, that makes sense. I see two issue:
> > > 
> > > 1. the "id" is the request is generated by Kodi. I would make your compare on just the `params` instead of the full `request`.
> > > 2. There is still no way to just _ignore_ the `request` and return a specific response based on just the json RPC method name. For example:  I stub the `Textures.RemoveTexture` and always want the same response, regardless of the input. Can we add a catch-all the filtering of the responses? If no specific response is found, return the catch-all if there is one, else return the default on from code.
> > 
> > 
> > Funny, in my tests I always get id: 0 so I think the counter in saké is not working right or maybe not implemented. :)
> 
> I guess that is true. But still, any add-on could decide to make it > 0. The second item is more important, I think.

Added backwards compatibility, if the logic finds just a dict inside the file its retuned else it will go through the whole process of determining the filed based on method and params combination.

**Review by basrieter — APPROVED** [review]

**Comment by basrieter** [comment]
Thanks, I will push a new version to PyPI soon

File: xbmc.py [hunk 2]
```diff
@@ -317,15 +315,24 @@ def executeJSONRPC(jsonrpccommand):  # NOSONAR
...
-    json_response = os.path.join(json_responses, "{}.json".format(json_method.lower()))
+        raise ValueError(
+            "Could not find JSON Response folder. Use the environment variable KODI_STUB_RPC_RESPONSES to set one")
+
+    path = "{}.json".format(os.path.join(os.path.join(json_responses, json_method.lower())))
+    if os.path.isfile(path):
+        with io.open(path, mode='r', encoding='utf-8') as fd:
+            stub_content = json.loads(fd.read())
+            if isinstance(stub_content, dict):
+                return json_data.dumps(stub_content)
```
Comment (basrieter) [inline xbmc.py:326]: I actually found an issue. This should be `json.dumps()` will fix in the final version of 0.0.3

Comment (MarcelRoozekrans) [reply xbmc.py:326]: Whoops, seems that I made this fuck up. Actually did not do a test run with the backwards compatible logic.
